src/ripple/shamap/SHAMapItem.h

```diff
@@ -149,36 +149,42 @@ intrusive_ptr_release(SHAMapItem const* x)
         if constexpr (!std::is_trivially_destructible_v<SHAMapItem>)
             std::destroy_at(x);
 
-        // If the slabber doens't claim this pointer, it was allocated
+        // If the slabber doesn't claim this pointer, it was allocated
         // manually, so we free it manually.
         if (!detail::slabber.deallocate(const_cast<std::uint8_t*>(p)))
```

Comment:
I'll take this opportunity to ask you about the choice of `boost::intrusive_ptr`, although its unrelated to this PR.
Why didn't you use a `std::shared_ptr` instead? I know that `boost::intrusive_ptr` has a smaller memory footprint.

The `intrusive_ptr_release` appears to be non-trivial. Can it not be implemented with a `std::shared_ptr` ?

Reply:
The question "Can it not be implemented with a `std::shared_ptr` is not the right question to be asking. The right question is "is `std::shared_ptr` appropriate?

And in many cases (most, even!) it will be appropriate. But in this case it isn't.

Using `boost::intrusive_ptr` allows us to squeeze a lot of savings in terms of size. And because there are _so many_ instances of this class, those small savings in terms of size add up.

Even if that was not a concern, `std::shared_ptr` would impose other limitations as well:

- First note that the fact that the memory allocation performed by `std::make_shared` is out of our control (that is, it cannot use the slabbed allocator).
- Second, it's important to understand what `std::make_shared<SHAMapItem>(...)` does: it allocates a block of memory that is large enough for the `std::shared_ptr` control block _and_ at least `sizeof(SHAMapItem)` extra bytes. But that's not enough: we need more data for the `SHAMapItem` buffer and we need that data to be _contiguous_ to the `SHAMapItem` itself, and there's no mechanism to allow `std::make_shared` to do that either.

It's true that `intrusive_ptr_release` is not trivial, but it's also not overly complex either. It basically does what `std::shared_ptr` does, only that is abstracted away and hidden from your view. And that does matter in many cases, but in _this_ case, it doesn't matter enough, at least IMO.

Reply:
I see, okay.
Appropos the second bullet point: 
```
... But that's not enough: we need more data for the SHAMapItem buffer and we need that data to be contiguous to the SHAMapItem itself ...
```

Why do we need more contiguous buffer for the SHAMapItem? Is it required to store the next pointer of linked list of items? (as you explained elsewhere in the comments) ?

Reply:
No, it's required to store the "contents" of the `SHAMapItem`. As I explained, previously a `SHAMapItem` contained a `std::vector` or a `Buffer` as a member (don't remember which off-hand) and that object required a separate allocation.

The new code allocates the space for the `SHAMapItem` and its dynamically-sized contents in one go, so that the contents are located immediately after the `SHAMapItem` in memory. For more details see commit c3acbce82dc5db8765cfc9dfeae36954cec0312f and checkout the changes in `SHAMapItem.h` and `SHAMapItem.cpp`.